### PR TITLE
Scroll editor ro is

### PR DESCRIPTION
**Pull request by zmbq**
This pull requests contains the ROI marks on the scroll editor, the text on the scroll editor (in the wrong location...) and some fixes frontend errors with Vue properties.

### CODE REVIEW

File: src/components/artefact/artefact-image.vue [hunk 1]
```diff
@@ -2,13 +2,20 @@
...
+        <g v-if="!this.artefact.isVirtual">
+            <iiif-image 
+                v-for="imageSetting in visibleImageSettings"
+                :key="imageSetting.image.url"
+                :image="imageSetting.image"
+                :opacity="imageSetting.normalizedOpacity"
+                :boundingBox="boundingBox"
+                :maxWidth="maxWidth"/>
+        </g>
+        <path v-if="artefact.isVirtual"
```
Comment (Bronson-Brown-deVost) [inline src/components/artefact/artefact-image.vue:14]: You can also use `v-else`, but maybe the explicit check is somehow safer.

File: src/models/artefact.ts [hunk 2]
```diff
@@ -25,14 +25,24 @@ export class Artefact {
     constructor(obj: ArtefactDTO) {
         this.id = obj.id;
         this.editionId = obj.editionId;
-        this.imagedObjectId = obj.imagedObjectId;
+        this.imagedObjectId = obj.imagedObjectId || '';
```
Comment (Bronson-Brown-deVost) [inline src/models/artefact.ts:28]: Maybe `??` is better in this case?  Or maybe the TS/Babel version we use doesn't yet support it.

File: src/models/artefact.ts [hunk 3]
```diff
@@ -25,14 +25,24 @@ export class Artefact {
...
         this.mask = Polygon.fromWkt(obj.mask);
         this.artefactMaskEditorId = obj.artefactMaskEditorId;
         this.isPlaced = obj.isPlaced;
         this.placement = new Placement(obj.placement);
         this.artefactPlacementEditorId = obj.artefactPlacementEditorId;
         this.side = (obj.side === 'recto') ? 'recto' : 'verso';
+
+        // TODO: Check if the artefact ID is on of the list of "virtual" artefacts. If so, set their
+        // imageObjectId to '', so that they are treated as virtual
+        if (this.id === 16772) {
```
Comment (Bronson-Brown-deVost) [inline src/models/artefact.ts:39]: This doesn't belong in the PR, does it?  It looks like some debugging stuff.

File: src/state/state-service.ts [hunk 4]
```diff
@@ -273,30 +274,37 @@ export default class StateService {
             throw new Error(`Can't located artefact ${artefactId} in edition ${editionId}`);
         }
 
-        const imagedObject = this._state.imagedObjects.find(artefact.imagedObjectId);
-        if (!imagedObject) {
-            console.error(`Can't locate imaged object ${artefact.imagedObjectId} for artefact ${artefact.id}`);
-            throw new Error(`Can't locate imaged object ${artefact.imagedObjectId} for artefact ${artefact.id}`);
-        }
+        const promises: Array<Promise<any>> = [];
```
Comment (Bronson-Brown-deVost) [inline src/state/state-service.ts:277]: Looks like the type should be whatever `this.imageManifest()` returns, or did I miss reason why we need to use the type `any`?

File: src/state/state-service.ts [hunk 5]
```diff
@@ -273,30 +274,37 @@ export default class StateService {
...
+        let imagedObject: ImagedObject | undefined;
+        if (!artefact.isVirtual) {
+            imagedObject = this._state.imagedObjects.find(artefact.imagedObjectId);
+            if (!imagedObject) {
+                console.error(`Can't locate imaged object ${artefact.imagedObjectId} for artefact ${artefact.id}`);
+                throw new Error(`Can't locate imaged object ${artefact.imagedObjectId} for artefact ${artefact.id}`);
```
Comment (Bronson-Brown-deVost) [inline src/state/state-service.ts:283]: Probably no need to throw an error, just run a GET on `v1/imaged-objects/{imagedObjectId}`, I doubt you would get an error then.

File: src/views/artefact-editor/artefact-editor.vue [hunk 6]
```diff
@@ -509,25 +508,32 @@ export default class ArtefactEditor
...
-            throw new Error(
-                `ImagedObject ${this.artefact.imagedObjectId} doesn't contain the ` +
-                    `${this.artefact.side} side even though artefact ${this.artefact.id} references it`
+
+        if (!this.artefact.isVirtual) {
+            const imagedObject = this.$state.imagedObjects.find(
+                this.artefact.imagedObjectId
             );
+            if (!imagedObject) {
+                throw new Error(
```
Comment (Bronson-Brown-deVost) [inline src/views/artefact-editor/artefact-editor.vue:517]: See above.

File: src/views/imaged-object-editor/imaged-object-editor-toolbar.vue [hunk 7]
```diff
@@ -318,45 +315,42 @@ export default class ImagedObjectEditorToolbar extends Vue {
         this.params.rotationAngle += degrees;
         this.notifyChange('rotationAngle', this.params.rotationAngle);
     }
-    public async newArtefact() {
-        this.newArtefactName = this.newArtefactName.trim();
+    // public async newArtefact() {
```
Comment (Bronson-Brown-deVost) [inline src/views/imaged-object-editor/imaged-object-editor-toolbar.vue:318]: Kill the code if you don't need it.

File: src/views/scroll-editor/artefact-image-group.vue [hunk 8]
```diff
@@ -14,24 +14,57 @@
...
         <g @click="onSelect">
-            <g :clip-path="`url(#clip-path-${artefact.id})`">
+            <g :clip-path="`url(#clip-path-${artefact.id})`" v-if="!artefact.isVirtual">
                 <iiif-image
                     :image="masterImage"
                     :boundingBox="boundingBox"
                     :scaleFactor="scaleFactor"
                 />
             </g>
+            <path v-if="artefact.isVirtual"
```
Comment (Bronson-Brown-deVost) [inline src/views/scroll-editor/artefact-image-group.vue:33]: Something looks wrong to me here.  In this path element and the one below it you check if `artefact.isVirtual`, but both elements are exactly the same except for one being of class `virtual-artefact` and the other being `virtual`.  Is there any reason you need two elements?

**Review by Bronson-Brown-deVost — CHANGES_REQUESTED** [review]
Please take a look at the few questions I had, sorry for any mistakes on my part.  In general it looks good to me.

File: src/components/artefact/artefact-image.vue (hunk 1 above)
Comment (zmbq) [reply src/components/artefact/artefact-image.vue:14]: I don't like `v-else` much, since it can break if I add another condition somewhere. Since ifs are not really nested, but rather just attributes, it can be hard to catch.

File: src/models/artefact.ts (hunk 2 above)
Comment (zmbq) [reply src/models/artefact.ts:28]: In principal you are right, but we do have a lot of `||` used just for that (because `??` was not introduced into Typescript when we started working on the project), so I don't think it matters much.

File: src/models/artefact.ts (hunk 3 above)
Comment (zmbq) [reply src/models/artefact.ts:39]: Oh, right! We should take this out!

File: src/state/state-service.ts (hunk 4 above)
Comment (zmbq) [reply src/state/state-service.ts:277]: We did not create a typescript type for the image manifest, so imageManifest is an any. We have an interface for the tile info, I guess we should create an interface for the parts of the manifests we use. I don't want this to delay progress in more crucial issues.

File: src/state/state-service.ts (hunk 5 above)
Comment (zmbq) [reply src/state/state-service.ts:283]: We definitely need to throw an error here - it means the store is in an inconsistent state. When this code runs, the imaged object should have been preloaded into the state already.

File: src/views/artefact-editor/artefact-editor.vue (hunk 6 above)
Comment (zmbq) [reply src/views/artefact-editor/artefact-editor.vue:517]: We have the `prepare` methods of in our internal state, that should make sure we have this already. If we don't, something is really messed up.

File: src/views/imaged-object-editor/imaged-object-editor-toolbar.vue (hunk 7 above)
Comment (zmbq) [reply src/views/imaged-object-editor/imaged-object-editor-toolbar.vue:318]: Right.

File: src/components/artefact/artefact-image.vue (hunk 1 above)
Comment (Bronson-Brown-deVost) [reply src/components/artefact/artefact-image.vue:14]: That sounds fair to me.

File: src/views/scroll-editor/artefact-image-group.vue (hunk 8 above)
Comment (zmbq) [reply src/views/scroll-editor/artefact-image-group.vue:33]: I don't see why we need two. I'll confirm with Shaindel. Good catch!

File: src/models/artefact.ts (hunk 2 above)
Comment (Bronson-Brown-deVost) [reply src/models/artefact.ts:28]: That is fine too.  If || maintains consistency with the existing code base, that is fine, I will try to keep an eye out for the possible unintended evaluations.

File: src/state/state-service.ts (hunk 4 above)
Comment (Bronson-Brown-deVost) [reply src/state/state-service.ts:277]: That is fine.

File: src/state/state-service.ts (hunk 5 above)
Comment (Bronson-Brown-deVost) [reply src/state/state-service.ts:283]: Is there a way to recover from this, or do you need to ask the user to reload?

**Review by Bronson-Brown-deVost — APPROVED** [review]
Many thanks.
